README.md
---
title: SLOANE(1) Sloane User Manual | Version 1.8
date: May 27, 2014
---

# NAME

sloane - a command line interface to Sloane's
On-Line Encyclopedia of Integer Sequences <http://oeis.org>

# SYNOPSIS

sloane [-a | --all | -k *keys* | --url] [-n *entries*] *search-terms* ...  
sloane (--update | --version | --help)  
sloane

# DESCRIPTION

The sloane command searches Sloane's On-Line Encyclopedia of Integer
Sequences (OEIS). The search terms are typically the leading term of a
sequence. For example,

    sloane 1,1,2,5,15,52,203,877,4140

returns entry A000110 (Bell numbers), and four more entries. One can
also search by sequence id (A-number), or even search for arbitrary
words. See the **EXAMPLES** section.

If no search terms are specified the standard input is read
line-by-line. In this mode the search is done locally against a
downloaded list of known sequences. If the sequence is in the OEIS, then
it is returned to the standard output; if not, it is ignored. This way
sloane can quickly filter out the sequences from the input that are in
the OEIS. Assuming that *FILE* contains one sequence per line,

    sloane <FILE

returns the subset of the sequences in *FILE* that are in the OEIS. To
also look-up the names of those sequences in the OEIS one could, for
instance, run

    sloane <FILE | xargs -L1 --verbose sloane

Sloane normally crops long lines to fit the widths of the terminal. If
this is unwanted, pipe the output through cat.

# OPTIONS

-a, --all
:   Print all fields

-k *keys*
:   Keys of fields to print (default: SN)

--url
:   Print URLs of found entries (but nothing else)

-n *entries*
:   Fetch at most this many entries (default: 5)

--update
:   Update the local sequence cache

--version
:   Print version information

--help
:   Display a short help message

# EXAMPLES

The most common search is for entries matching a sequence of consecutive terms:

    sloane 1,3,19,183,2371,38703

At the time of writing this particular query would return

    S A006531 1,1,3,19,183,2371,38703,763099,17648823,468603091,14050842303,
    N A006531 Semiorders on n elements.

As this illustrates, the default is to return just the sequence (S) and
the name (N) fields. To override the default one can use the keys
option. For instance, the following search shows the sequence, name,
comments, and formula fields of the sequence whose A-number is A006531:

    sloane -k SNCF id:A006531

The next example returns at most 3 results of a free text search:

    sloane -n 3 "(2+2)-free posets"

To view the full entries of these 3 results in a browser (e.g., Firefox)
one can use the url option:

    firefox `sloane --url -n 3 "(2+2)-free posets"`

In the final example the local cache is used to filter out sequences
from the standard input that are in the OEIS:

    sloane <<END
    1,2,3,6,11,23,47,106,235           # Comma separated integers
    1 2 444 90 120                     # Space separated integers
    '(3 9 27 88 123)                   # S-expression
    [2,3,5,7,11,13,17,19,23,29,31,37]  # Haskell list
    [1; 1; 2; 3; 5; 8; 13; 21]         # O'Caml list
    {1, -1, 2, -6, 24, -120, 720}      # Mathematica list
    END

The '#' character and any text after it is ignored by sloane. Note also
that sloane is quite liberal in the formatting of sequences it accepts.

# KEYS

These are the keys used by OEIS <http://oeis.org/eishelp2.html>.

    I  ID number

    S  1st line of unsigned sequence
    T  2nd line of unsigned sequence
    U  3rd line of unsigned sequence

    V  1st line of signed sequence
    W  2nd line of signed sequence
    X  3rd line of signed sequence

    N  Name
    C  Comments
    D  References
    H  Links
    F  Formula
    e  Examples

    p  Maple program
    t  Mathematica program
    o  Program in other language

    Y  Cross-references
    K  Keywords
    O  Offset
    A  Author
    E  Extensions and errors

# NOTES

Please use this program with moderation as not to overburden the
OEIS-server; see OEIS' policy on searching the database:
<http://oeis.org/wiki/Welcome#Policy_on_Searching_the_Database>.

# SEE ALSO

The sloane source code may be downloaded from
<https://github.com/akc/sloane>.

# AUTHOR

Anders Claesson <http://akc.is>


CHANGELOG.md
# Changelog

## sloane 1.8

Use optparse-applicative instead of cmdargs for command line parsing.
Deprecate '-u', '-?', '-V', '--limit', and '--keys'.

## sloane 1.7.1 - April 30, 2014

* Clean up the README / man page.

## sloane 1.7 - April 20, 2014

* Enable filtering against a local cache of sequences.

## sloane 1.6 - March 18, 2014

* Properly handle multiple search terms.


sloane.hs
-- |
-- Copyright   : Anders Claesson 2012-2014
-- Maintainer  : Anders Claesson <anders.claesson@gmail.com>
-- License     : BSD-3
--
import qualified Codec.Compression.GZip       as GZip
import           Control.Monad                (unless, when)
import qualified Data.ByteString.Char8        as B
import qualified Data.ByteString.Lazy.Char8   as BL
import qualified Data.ByteString.Lazy.Search  as Search
import           Data.Maybe                   (fromJust, maybe)
import           Data.Time                    (diffUTCTime, getCurrentTime)
import           Network.HTTP
import           Network.URI                  (parseURI)
import           Options.Applicative
import           System.Console.ANSI
import           System.Console.Terminal.Size (Window (..), size)
import           System.Directory
import           System.FilePath              ((</>))
import           System.IO                    (stderr, stdin)

type OEISEntries = [String]
type ANumbers = [String]
type Query = String
type Keys = String

data Args = Args
    { all    :: Bool
    , keys   :: String
    , limit  :: Int
    , update :: Bool
    , url    :: Bool
    , ver    :: Bool
    , terms  :: [String]
    }

version   = "sloane 1.8"

oeisHost  = "http://oeis.org/"
oeisURL   = oeisHost ++ "search?fmt=text"
oeisKeys  = "ISTUVWXNDHFYAOEeptoKC"

cacheDir  = ".sloane"
cacheFile = "stripped.gz"
cacheURL  = oeisHost ++ cacheFile

msgDownloadingCache = unlines
    [ "Downloading " ++ cacheURL
    , "This may take a minute or two ..."
    ]
msgCacheIsUpToDate = unlines
    [ "The sequence cache is now up-to-date"
    ]
msgNoCache = unlines
    [ "No sequence cache found. You need to run \"sloane --update\""
    ]
msgOldCache = unlines
    [ "The sequence cache is more than 100 days old"
    , "You may want to run \"sloane --update\""
    ]

select :: Keys -> OEISEntries -> OEISEntries
select ks = filter (\line -> null line || head line `elem` ks)

aNumbers :: OEISEntries -> ANumbers
aNumbers es = [ words ids !! 1 | ids@(_:_) <- select "I" es ]

urls :: OEISEntries -> String
urls = unlines . map (oeisHost ++ ) . aNumbers

get :: HStream b => String -> IO b
get uri = simpleHTTP (defaultGETRequest_ uri') >>= getResponseBody
  where
    uri' = fromJust $ parseURI uri

updateCache :: FilePath -> IO ()
updateCache home = do
    let dir = home </> cacheDir
    createDirectoryIfMissing False dir
    put msgDownloadingCache
    cache <- get cacheURL
    B.writeFile (dir </> cacheFile) cache
    put msgCacheIsUpToDate

searchOEIS :: Int -> Query -> IO OEISEntries
searchOEIS n s = trim `fmap` get uri
  where
    trim = map (drop 1) . reverse . drop 2 . reverse . drop 5 . lines
    uri = oeisURL ++ "&" ++ urlEncodeVars [("n", show n), ("q", s)]

cropStr :: (Int -> String -> String) -> Int -> String -> String
cropStr f maxLen s = if maxLen < length s then f maxLen s else s

cropSeq :: Int -> String -> String
cropSeq = cropStr $ \maxLen ->
              reverse . dropWhile (/= ',') . reverse . take maxLen

cropLine :: Int -> String -> String
cropLine = cropStr $ \maxLen s -> take (maxLen-2) s ++ ".."

getWidth :: IO Int
getWidth = maybe maxBound width `fmap` size

put = B.putStr . B.pack
putErr = B.hPutStr stderr . B.pack
newline = B.putStrLn B.empty

putEntries :: Int -> OEISEntries -> IO ()
putEntries width = mapM_ $ \line ->
    case words line of
        [] -> newline
        (key:aNum:rest) -> do
            setSGR [ SetColor Foreground Dull Green ]
            put key
            setSGR [ SetColor Foreground Dull Yellow ]
            put $ ' ' : aNum
            setSGR []
            let crop = if key == "S" then cropSeq else cropLine
            put $ ' ' : crop width (unwords rest) ++ "\n"

isInfixOf :: B.ByteString -> BL.ByteString -> Bool
isInfixOf q = not . null . Search.indices q

dropPreamble :: BL.ByteString -> BL.ByteString
dropPreamble = BL.unlines . drop 4 . BL.lines

readCache :: FilePath -> IO BL.ByteString
readCache home = do
    let name = home </> cacheDir </> cacheFile
    updated <- doesFileExist name
    if updated
        then do
            c <- getCurrentTime
            m <- getModificationTime name
            let day = 60*60*24
            let expired = c `diffUTCTime` m > 100*day
            when expired $ putErr msgOldCache
            (dropPreamble . GZip.decompress) `fmap` BL.readFile name
        else
            error msgNoCache

seqs :: B.ByteString -> [B.ByteString]
seqs = filter (not . B.null) . map mkSeq . B.lines
  where
    mkSeq = normalize . dropComment
    dropComment = B.takeWhile (/= '#')
    normalize   = B.intercalate (B.pack ",") . B.words . clean . B.map tr
    tr c  = if c `elem` ";," then ' ' else c
    clean = B.filter (\c -> B.elem c (B.pack " 0123456789-"))

filterSeqs :: FilePath -> IO ()
filterSeqs home = do
    cache <- readCache home
    B.getContents >>= mapM_ B.putStrLn . filter (`isInfixOf` cache) . seqs

args :: Parser Args
args = Args
    <$> switch (short 'a' <> long "all" <> help "Print all fields")
    <*> strOption
        ( short 'k'
       <> metavar "KEYS"
       <> value "SN"
       <> help "Keys of fields to print [default: SN]" )
    <*> option
        ( short 'n'
       <> metavar "N"
       <> value 5
       <> help "Fetch at most this many entries [default: 5]" )
    <*> switch (long "update" <> help "Update the local sequence cache")
    <*> switch (long "url" <> help "Print URLs of found entries")
    <*> switch (hidden <> long "version")
    <*> many (argument str (metavar "TERMS..."))

sloane :: Args -> IO ()
sloane (Args all keys n update url True terms) = put version >> newline
sloane (Args all keys n True   url ver  terms) = getHomeDirectory >>= updateCache
sloane (Args all keys n update url ver  []   ) = getHomeDirectory >>= filterSeqs
sloane (Args all keys n update url ver  terms) = do
    ncols <- getWidth
    hits  <- searchOEIS n (unwords terms)
    let pick = if all then id else select keys
    unless (null hits) $ do
        newline
        if url
            then put (urls hits)
            else putEntries (ncols - 10) (pick hits)
        newline

main = execParser (info (h <*> args) (fullDesc <> header version)) >>= sloane
  where
    h = abortOption ShowHelpText $ hidden <> long "help"
